### PR TITLE
ois based Intra precition

### PR DESCRIPTION
**Pull request by okhlif**
Upgrade OIS to AV1
OIS based intra prediction in M7 and higher
Changed Intra prediction settings for M0 and M1

### CODE REVIEW

**Comment by jbeich** [comment]
Appears to break `-enc-mode` < 3 in some cases e.g.,

```
$ curl -sO https://media.xiph.org/video/derf/y4m/elephants_dream_720p24.y4m.xz
$ xz -d elephants_dream_720p24.y4m.xz
$ ffmpeg -hide_banner -i elephants_dream_720p24.y4m -frames:v 120 -y ed.yuv

$ SvtAv1EncApp -enc-mode 1 -i ed.yuv -w 1280 -h 720 -fps 24 -b ed.ivf
-------------------------------------------
SVT-AV1 Encoder
SVT [version]:  SVT-AV1 Encoder Lib v0.4.0
SVT [build]  :  GCC 4.2.1        64 bit
LIB Build date: Mar  6 2019 18:37:17 (dbabacee0548)
-------------------------------------------
Number of logical cores available: 8
Number of PPCS 41
-------------------------------------------
SVT [config]: Main Profile      Tier (auto)     Level (auto)
SVT [config]: EncoderMode                                                       : 1
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1280 / 720
SVT [config]: FrameRate / Gop Size                                              : 24 / 32
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 4 / 0 / 2
SVT [config]: BRC Mode / QP  / LookaheadDistance / SceneChange                  : CQP / 50 / 33 / 0
-------------------------------------------
Encoding          

Thread 19 received signal SIGSEGV, Segmentation fault.
[Switching to LWP 101632 of process 25465]
0x00000008014ed1a4 in open_loop_intra_search_sb (picture_control_set_ptr=0x85e2b4c80, sb_index=0,
    context_ptr=0x8f9ad50f0, input_ptr=0x8dec4d3a0, asm_type=ASM_AVX2)
    at Source/Lib/Codec/EbMotionEstimation.c:8263
8263                                ois_blk_ptr[ois_intra_count].distortion = (uint32_t)NxMSadKernel_funcPtrArray[asm_type][blk_geom->bwidth >> 3]( // Always SAD without weighting
(gdb) backtrace full
#0  0x00000008014ed1a4 in open_loop_intra_search_sb (picture_control_set_ptr=0x85e2b4c80, sb_index=0,
    context_ptr=0x8f9ad50f0, input_ptr=0x8dec4d3a0, asm_type=ASM_AVX2)
    at Source/Lib/Codec/EbMotionEstimation.c:8263
        left_col = 0x7fffdd7e9c70 '\020' <repeats 128 times>
        above_data = '\000' <repeats 15 times>, '\020' <repeats 33 times>, '\177' <repeats 96 times>, '\000' <repeats 15 times>
        best_intra_ois_index = 0 '\000'
        intra_mode_end = 12 '\f'
        angle_delta_counter = 0 '\000'
        ois_blk_ptr = 0x0
        left_data = '\000' <repeats 15 times>, '\020' <repeats 129 times>, '\000' <repeats 15 times>
        ois_intra_mode = 0 '\000'
        best_intra_ois_distortion = 1044480
        above_row = 0x7fffdd7e9bd0 '\020' <repeats 32 times>, '\177' <repeats 96 times>
        ois_intra_count = 0 '\000'
        intra_mode_start = 0 '\000'
        use_angle_delta = 1 '\001'
        angle_delta_candidate_count = 5 '\005'
        disable_angular_prediction = 0 '\000'
        blk_geom = 0x801a4f544 <blk_geom_mds+143428>
        return_error = EB_ErrorNone
        sequence_control_set_ptr = 0x85aa7e600
        cu_origin_x = 96
        cu_origin_y = 16
        ep_blk_index = 1559
        pa_blk_index = 12336
        next_sq = 1 '\001'
        is_16_bit = 0 '\000'
        sb_params = 0x91dc004c0
        ois_sb_results_ptr = 0x801110380
        top_neigh_array = '\020' <repeats 33 times>, '\177' <repeats 96 times>
        left_neigh_array = '\020' <repeats 129 times>
        above_ref = 0x7fffdd7e9d90 '\020' <repeats 33 times>, '\177' <repeats 167 times>...
        left_ref = 0x7fffdd7e9d00 '\020' <repeats 129 times>, '\201' <repeats 15 times>, '\020' <repeats 33 times>, '\177' <repeats 23 times>...
#1  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#2  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#3  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#4  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#5  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#6  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#7  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#8  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#9  0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#10 0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#11 0x7f7f7f7f7f7f7f7f in ?? ()
No symbol table info available.
#12 0x0000000a0000007f in ?? ()
No symbol table info available.
#13 0x000000400000000a in ?? ()
No symbol table info available.
#14 0x0000000000000000 in ?? ()
No symbol table info available.

=================================================================
==29504==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fffdd7e8c21 at pc 0x00000107bc82 bp 0x7fffdd7e8a00 sp 0x7fffdd7e81b0
WRITE of size 257 at 0x7fffdd7e8c21 thread T18
    #0 0x107bc81 in __asan_memset (/usr/local/bin/SvtAv1EncApp+0x5ac81)
    #1 0x8017344cf in update_neighbor_samples_array_open_loop Source/Lib/Codec/EbIntraPrediction.c:10166:5
    #2 0x801815681 in open_loop_intra_search_sb Source/Lib/Codec/EbMotionEstimation.c:8176:17
    #3 0x801837648 in MotionEstimationKernel Source/Lib/Codec/EbMotionEstimationProcess.c:570:21
    #4 0x8011ac775 in thread_start /usr/src/lib/libthr/thread/thr_create.c:291:16

Address 0x7fffdd7e8c21 is located in stack of thread T18 at offset 161 in frame
    #0 0x801814c0f in open_loop_intra_search_sb Source/Lib/Codec/EbMotionEstimation.c:8145

  This frame has 4 object(s):
    [32, 161) 'top_neigh_array' (line 8159)
    [240, 369) 'left_neigh_array' (line 8160) <== Memory access at offset 161 partially underflows this variable
    [448, 608) 'left_data' (line 8189)
    [672, 832) 'above_data' (line 8190)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
Thread T18 created by T0 here:
    #0 0x10fdabd in __interceptor_pthread_create (/usr/local/bin/SvtAv1EncApp+0xdcabd)
    #1 0x801a4c9c8 in eb_create_thread Source/Lib/Codec/EbThreads.c:95:13
    #2 0x8015f0d0d in eb_init_encoder Source/Lib/Codec/EbEncHandle.c:1836:9
    #3 0x113c81e in InitEncoder Source/App/EbAppContext.c:615:20
    #4 0x114081b in main Source/App/EbAppMain.c:161:52
    #5 0x107411c in _start /usr/src/lib/csu/amd64/crt1.c:76:7
    #6 0x80117ffff  (<unknown module>)

SUMMARY: AddressSanitizer: stack-buffer-overflow (/usr/local/bin/SvtAv1EncApp+0x5ac81) in __asan_memset
Shadow bytes around the buggy address:
  0x4ffffbafd130: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x4ffffbafd140: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x4ffffbafd150: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x4ffffbafd160: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x4ffffbafd170: f1 f1 f1 f1 00 00 00 00 00 00 00 00 00 00 00 00
=>0x4ffffbafd180: 00 00 00 00[01]f2 f2 f2 f2 f2 f2 f2 f2 f2 00 00
  0x4ffffbafd190: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 f2
  0x4ffffbafd1a0: f2 f2 f2 f2 f2 f2 f2 f2 f8 f8 f8 f8 f8 f8 f8 f8
  0x4ffffbafd1b0: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2
  0x4ffffbafd1c0: f2 f2 f2 f2 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x4ffffbafd1d0: f8 f8 f8 f8 f8 f8 f8 f8 f3 f3 f3 f3 f3 f3 f3 f3
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==29504==ABORTING
```

**Comment by okhlif** [comment]
Addressed , Thank you

File: Source/Lib/Codec/EbInitialRateControlProcess.c [hunk 1]
```diff
@@ -1359,6 +1359,15 @@ void QpmGatherStatisticsSW(
             if (sb_params.raster_scan_cu_validity[rasterScanCuIndex]) {
                 mdScanCuIndex = RASTER_SCAN_TO_MD_SCAN[rasterScanCuIndex];
 
+#if OIS_BASED_INTRA
+
+
+            ois_sb_results_t        *ois_sb_results_ptr = picture_control_set_ptr->ois_sb_results[sb_index];
+	
+            ois_candidate_t *OisCuPtr = ois_sb_results_ptr->ois_candidate_array[mdScanCuIndex];
```
Comment (joelsole) [inline Source/Lib/Codec/EbInitialRateControlProcess.c]: snake case

File: Source/Lib/Codec/EbInitialRateControlProcess.c [hunk 2]
```diff
@@ -1405,12 +1414,19 @@ void QpmGatherStatisticsSW(
     cu_depth = 2;
     for (rasterScanCuIndex = RASTER_SCAN_CU_INDEX_16x16_0; rasterScanCuIndex <= RASTER_SCAN_CU_INDEX_16x16_15; rasterScanCuIndex++) {
         if (sb_params.raster_scan_cu_validity[rasterScanCuIndex]) {
+#if OIS_BASED_INTRA
+            
+            mdScanCuIndex = RASTER_SCAN_TO_MD_SCAN[rasterScanCuIndex];
 
+            ois_sb_results_t        *ois_sb_results_ptr = picture_control_set_ptr->ois_sb_results[sb_index];	
+            ois_candidate_t *OisCuPtr = ois_sb_results_ptr->ois_candidate_array[mdScanCuIndex];
```
Comment (joelsole) [inline Source/Lib/Codec/EbInitialRateControlProcess.c]: snake case

File: Source/Lib/Codec/EbInitialRateControlProcess.c (hunk 1 above)
Comment (okhlif) [reply Source/Lib/Codec/EbInitialRateControlProcess.c]: Fixed Thank you

File: Source/Lib/Codec/EbInitialRateControlProcess.c (hunk 2 above)
Comment (okhlif) [reply Source/Lib/Codec/EbInitialRateControlProcess.c]: Fixed Thank you

**Comment by hassount** [comment]
Included in #185

**Pull request closed without merging**